yunojuno/candidate-test-24465216
# Grant or deny access based on number of visits associated with a token.

**What does this PR do?**

- Add field `max_visits_allowed` on the `Visitor` model to specify the maximum number of allowed visits by a visitor. If this value is set to `null` or to `0` tracking access based on allowed number of visits is not considered, unless this is set to a value greater or equal to `1`. Default value is `null`.
- Add field `num_of_visits` on the `Visitor` model to track the number of visits for a visitor.
- Add model method `track_num_of_visits` on the `Visitor` model to increment the `num_of_visits` by 1. This method will be available to be manually triggered for use cases when a view does not use the `user_is_visitor` decorator.
- Trigger model method `track_num_of_visits` in the `user_is_visitor` decorator in order to increment `num_of_visits` when a visitor accesses the decorated view successfully.
- Update VisitorSessionMiddleware middleware to clear visitor session data when the allowed visits of a visitor have been exceeded.
- Update `validate` Visitor model method to raise an exception when allowed visits have been exceeded for a visitor.
- Update readme to show how to trigger `track_num_of_visits` Visitor model method for views that require visitors but don't use the `user_is_visitor` decorator.

**How should this be manually tested?**

- Pull changes from this branch to your local project.
- Make sure to run migrations using `python manage.py migrate`.
- Run the development server `python manage.py runserver`.
- Access the django admin at path `/admin`.
- Add a new `Visitor Pass` entry in the django admin, make sure to set the `Max visits allowed` field to a value greater than `0` such as `2`. If it's left empty or set to `0` access based on visits by the visitor will not be considered.
- Visit the endpoint `/foo/?vuid=<token>` using the token generated from above you should see a response of `OK` until you visit the endpoint up to number specified in `Max visits allowed` field, then you should see a `403 Forbidden` response.
